In [0]:
import tensorflow as tf
tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, Activation, BatchNormalization
from keras.utils import np_utils
from keras.datasets import cifar10
from tensorflow.keras.optimizers import RMSprop
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten
from tensorflow.keras import regularizers

https://machinelearningmastery.com/how-to-develop-a-cnn-from-scratch-for-cifar-10-photo-classification/

In [3]:
# the data, shuffled and split between train and val sets
# Here we are using the official test set as our validation set, in further
# tutorials, test and validation splits will be explained properly.
(x_train, y_train), (x_test, y_test) = cifar10.load_data()

print('Image shape: {0}'.format(x_train.shape[1:]))
print('Total number of training samples: {0}'.format(x_train.shape[0]))
print('Total number of test samples: {0}'.format(x_test.shape[0]))

x_train = x_train.astype('float32')
x_test = x_test.astype('float32')

# Normalize the image
x_train /= 255
x_test /= 255

print(x_train.shape)

y_train_class = np_utils.to_categorical(y_train, 10)
y_test_class = np_utils.to_categorical(y_test, 10)

170500096/170498071 [==============================] - 11s 0us/step
Image shape: (32, 32, 3)
Total number of training samples: 50000
Total number of test samples: 10000
(50000, 32, 32, 3)


In [4]:
# Define here your architecture

model = Sequential()
#####################################################################
model.add(Conv2D(32, (3,3), padding='same', input_shape=(32, 32, 3)))
model.add(Activation('relu'))
model.add(Conv2D(32, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.2))
######################################################################
model.add(Conv2D(64, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(Conv2D(64, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.2))
######################################################################
model.add(Conv2D(128, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(Conv2D(128, (3,3), padding='same'))
model.add(Activation('relu'))
model.add(MaxPooling2D(pool_size=(2,2)))
model.add(Dropout(0.2))
######################################################################
model.add(Dense(128, activation='relu'))
model.add(Dropout(0.2))
######################################################################
model.add(Flatten())
model.add(Dense(10))
model.add(Activation('softmax'))
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 32, 32, 32)        896       
_________________________________________________________________
activation (Activation)      (None, 32, 32, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 32, 32, 32)        9248      
_________________________________________________________________
activation_1 (Activation)    (None, 32, 32, 32)        0         
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 16, 16, 32)        0         
_________________________________________________________________
dropout (Dropout)            (None, 16, 16, 32)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 16, 16, 64)        1

In [5]:
from tensorflow.keras.callbacks import EarlyStopping

# Define the metric that will determine the earlyStopping
# patience defines the number of epochs we wait before stopping training 
# when the monitored metric does not improve
early_stopping = EarlyStopping(monitor='val_loss', patience=5)

# initiate RMSprop optimizer
opt = RMSprop(lr=0.0001, decay=1e-6)

model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])

history = model.fit(x_train, y_train_class,validation_split=0.2, batch_size=32, epochs=50,callbacks=[early_stopping])

score = model.evaluate(x_test, y_test_class, verbose=0)
print('Test loss:', score[0])
print('Test accuracy:', score[1])

Epoch 1/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.9521 - accuracy: 0.2761 - val_loss: 1.7534 - val_accuracy: 0.3676
Epoch 2/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.6456 - accuracy: 0.3959 - val_loss: 1.5405 - val_accuracy: 0.4374
Epoch 3/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.5037 - accuracy: 0.4506 - val_loss: 1.4074 - val_accuracy: 0.4944
Epoch 4/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.3917 - accuracy: 0.4956 - val_loss: 1.2870 - val_accuracy: 0.5392
Epoch 5/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.2971 - accuracy: 0.5350 - val_loss: 1.2197 - val_accuracy: 0.5730
Epoch 6/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.2213 - accuracy: 0.5625 - val_loss: 1.1435 - val_accuracy: 0.5951
Epoch 7/50
1250/1250 [==============================] - 7s 6ms/step - loss: 1.1559 - accuracy: 0.5898 - val_loss: 1.1645 - val_accuracy:

In [16]:
(x_train1, y_train1), (x_test1, y_test1) = cifar10.load_data()

print('Image shape: {0}'.format(x_train1.shape[1:]))
print('Total number of training samples: {0}'.format(x_train1.shape[0]))
print('Total number of test samples: {0}'.format(x_test1.shape[0]))

x_train1 = x_train1.astype('float32')
x_test1 = x_test1.astype('float32')

# Normalize the image
x_train1 /= 255
x_test1 /= 255

Image shape: (32, 32, 3)
Total number of training samples: 50000
Total number of test samples: 10000


In [0]:
from keras.datasets import mnist
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout
from keras.utils.np_utils import to_categorical
from tensorflow.keras.callbacks import EarlyStopping
from tensorflow.keras.optimizers import RMSprop
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten
from tensorflow.keras import regularizers

# Helper: Early stopping.
early_stopper = EarlyStopping(patience=5)

def get_mnist():
    """Retrieve the MNIST dataset and process the data."""
    # Set defaults.
    nb_classes = 10
    batch_size = 128
    input_shape = x_train1.shape[1:]

    # Get the data.
    x_train = x_train1
    x_test = x_test1

    # convert class vectors to binary class matrices
    y_train = to_categorical(y_train1, nb_classes)
    y_test = to_categorical(y_test1, nb_classes)
    print('done mnist')
    return (nb_classes, batch_size, input_shape, x_train, x_test, y_train, y_test)

def compile_model(network, nb_classes, input_shape):
    """Compile a sequential model.
    Args:
        network (dict): the parameters of the network
    Returns:
        a compiled network.
    """
    # Get our network parameters.
        #'VGG_layers': range(1,5),
        #VGG_sublayers': range(1,4),

    VGG_layers = network['VGG_layers']
    VGG_sublayers = network['VGG_sublayers']
    pool_num =network['pool_num']

    nb_layers = network['nb_layers']
    nb_neurons1 = network['nb_neurons1']
    nb_neurons2 = network['nb_neurons2']
    nb_neurons3 = network['nb_neurons3']
    nb_neurons4 = network['nb_neurons4']
    activation = network['activation']
    optimizer = network['optimizer']

    model = Sequential()

    if VGG_layers == 0:                        #no conv layers
      for i in range(nb_layers):

          # Need input shape for first layer.
          if i == 0:
              model.add(Dense(nb_neurons1, activation=activation, input_shape=input_shape))
          elif i == 1:
              model.add(Dense(nb_neurons2, activation=activation))
          elif i == 2:
              model.add(Dense(nb_neurons3, activation=activation))
          elif i == 3:
              model.add(Dense(nb_neurons4, activation=activation))

          model.add(Dropout(0.2))  # hard-coded dropout

    else:
      for i in range(VGG_layers):
          if i == 0:
              model.add(Conv2D(32, (3,3), padding='same', input_shape=(32, 32, 3)))
              model.add(Activation(activation))
              if (VGG_sublayers > 1) :
                for x in range(1, VGG_sublayers):
                  model.add(Conv2D(32, (3,3), padding='same'))
                  model.add(Activation(activation))
              model.add(MaxPooling2D(pool_size=(pool_num,pool_num)))
          else:
              for x in range(0, VGG_sublayers):
                model.add(Conv2D(((i+1)*32), (3,3), padding='same'))
                model.add(Activation(activation))
              model.add(MaxPooling2D(pool_size=(pool_num,pool_num)))

      for i in range(nb_layers):
          # Need input shape for first layer.
          if i == 0:
              model.add(Dense(nb_neurons1, activation=activation))
          elif i == 1:
              model.add(Dense(nb_neurons2, activation=activation))
          elif i == 2:
              model.add(Dense(nb_neurons3, activation=activation))
          elif i == 3:
              model.add(Dense(nb_neurons4, activation=activation))

          model.add(Dropout(0.2))  # hard-coded dropout




    # Output layer.
    model.add(Flatten())
    model.add(Dense(nb_classes, activation='softmax'))

    model.compile(loss='categorical_crossentropy', optimizer=optimizer,
                  metrics=['accuracy'])
    print('compiled model')
    return model

def train_and_score(network, dataset):
    """Train the model, return test loss.
    Args:
        network (dict): the parameters of the network
        dataset (str): Dataset to use for training/evaluating
    """
    logging.info("***train_and_score***")
    if dataset == 'cifar10':
        nb_classes, batch_size, input_shape, x_train, \
            x_test, y_train, y_test = get_cifar10()
    elif dataset == 'mnist':
        logging.info("***mnist***")
        nb_classes, batch_size, input_shape, x_train, \
            x_test, y_train, y_test = get_mnist()

    model = compile_model(network, nb_classes, input_shape)
    logging.info("***compiled_model***")
    model.fit(x_train, y_train,
              batch_size=batch_size,
              epochs=100,  # using early stopping, so no real limit
              verbose=0,
              validation_split=0.2,
              callbacks=[early_stopper])
    #validation_split=0.2, epochs=100, batch_size=2000,
    #validation_data=(x_test, y_test),
    score = model.evaluate(x_test, y_test, verbose=0)
    logging.info("***trainedd***")
    print(score[1])
    return score[1]  # 1 is accuracy. 0 is loss.

Network

In [0]:
import random
import logging
#from train import train_and_score

class Network():
    """Represent a network and let us operate on it.
    Currently only works for an MLP.
    """

    def __init__(self, nn_param_choices=None):
        """Initialize our network.
        Args:
            nn_param_choices (dict): Parameters for the network, includes:
                nb_neurons (list): [64, 128, 256]
                nb_layers (list): [1, 2, 3, 4]
                activation (list): ['relu', 'elu']
                optimizer (list): ['rmsprop', 'adam']
        """
        self.accuracy = 0.
        self.nn_param_choices = nn_param_choices
        self.network = {}  # (dic): represents MLP network parameters

    def create_random(self):
        i = True 
        """Create a random network."""
        for key in self.nn_param_choices:
            if key == 'nb_neurons2':
              while (True):
                logging.info("stuck 1")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons1']:
                  break
            elif key == 'nb_neurons3':
              while (True):
                logging.info("stuck 2")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons2']:
                  break
            elif key == 'nb_neurons4':
              while (True):
                logging.info("stuck 3")
                self.network[key] = random.choice(self.nn_param_choices[key])
                if self.network[key] < self.network['nb_neurons3']:
                  break
            else:
              self.network[key] = random.choice(self.nn_param_choices[key])



    def create_set(self, network):
        """Set network properties.
        Args:
            network (dict): The network parameters
        """
        self.network = network

    def train(self, dataset):
        """Train the network and record the accuracy.
        Args:
            dataset (str): Name of dataset to use.
        """
        logging.info("***train***")
        if self.accuracy == 0.:
            self.accuracy = train_and_score(self.network, dataset)

    def print_network(self):
        """Print out a network."""
        logging.info(self.network)
        logging.info("Network accuracy: %.2f%%" % (self.accuracy * 100))

In [0]:
from functools import reduce
from operator import add
import random


class Optimizer():
    """Class that implements genetic algorithm for MLP optimization."""

    def __init__(self, nn_param_choices, retain=0.4,
                 random_select=0.1, mutate_chance=0.8):
        """Create an optimizer.
        Args:
            nn_param_choices (dict): Possible network paremters
            retain (float): Percentage of population to retain after
                each generation
            random_select (float): Probability of a rejected network
                remaining in the population
            mutate_chance (float): Probability a network will be
                randomly mutated
        """
        self.mutate_chance = mutate_chance
        self.random_select = random_select
        self.retain = retain
        self.nn_param_choices = nn_param_choices

    def create_population(self, count):
        """Create a population of random networks.
        Args:
            count (int): Number of networks to generate, aka the
                size of the population
        Returns:
            (list): Population of network objects
        """
        pop = []
        for _ in range(0, count):
            # Create a random network.
            network = Network(self.nn_param_choices)
            logging.info("count")
            network.create_random()

            # Add the network to our population.
            pop.append(network)

        return pop

    @staticmethod
    def fitness(network):
        """Return the accuracy, which is our fitness function."""
        return network.accuracy

    def grade(self, pop):
        """Find average fitness for a population.
        Args:
            pop (list): The population of networks
        Returns:
            (float): The average accuracy of the population
        """
        summed = reduce(add, (self.fitness(network) for network in pop))
        return summed / float((len(pop)))

    def breed(self, mother, father):
        """Make two children as parts of their parents.
        Args:
            mother (dict): Network parameters
            father (dict): Network parameters
        Returns:
            (list): Two network objects
        """
        children = []
        for _ in range(2):

            child = {}

            #  need to do if they have equal layers
            # Loop through the parameters and pick params for the kid.
            for param in self.nn_param_choices:
                  if (param == 'nb_neurons2') :
                    while (True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons1']:
                            logging.info("***broke1***")
                            break
                  elif param == 'nb_neurons3':
                    while ( True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons2']:
                            break
                            logging.info("***broke2***")
                  elif param == 'nb_neurons4':
                    while (True):
                        child[param] = random.choice([mother.network[param], father.network[param]])
                        if child[param] < child['nb_neurons3']:
                            break
                            logging.info("***broke3***")
                  else:
                    child[param] = random.choice([mother.network[param], father.network[param]])

            # Now create a network object.
            network = Network(self.nn_param_choices)
            network.create_set(child)

            # Randomly mutate some of the children.
            if self.mutate_chance > random.random():
                network = self.mutate(network)

            children.append(network)

        return children

    def mutate(self, network):
        """Randomly mutate one part of the network.
        Args:
            network (dict): The network parameters to mutate
        Returns:
            (Network): A randomly mutated network object
        """
        # Choose a random key.

 
        mutation = random.choice(list(self.nn_param_choices.keys()))
        
        #mutate param within condition
        if mutation == 'nb_neurons1':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons2']):
                    break
        elif mutation == 'nb_neurons2':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons3']) and (network.network[mutation] < network.network['nb_neurons1']) :
                    break
        elif mutation == 'nb_neurons3':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] > network.network['nb_neurons4'])  and (network.network[mutation] < network.network['nb_neurons2']):
                    break
        elif mutation == 'nb_neurons4':
            while (True):
                  network.network[mutation] = random.choice(self.nn_param_choices[mutation])
                  if (network.network[mutation] < network.network['nb_neurons3']):
                    break
        else:
          # Mutate one of the params.
          network.network[mutation] = random.choice(self.nn_param_choices[mutation])
        return network

    def evolve(self, pop):
        """Evolve a population of networks.
        Args:
            pop (list): A list of network parameters
        Returns:
            (list): The evolved population of networks
        """
        # Get scores for each network.
        graded = [(self.fitness(network), network) for network in pop]

        # Sort on the scores.
        graded = [x[1] for x in sorted(graded, key=lambda x: x[0], reverse=True)]

        # Get the number we want to keep for the next gen.
        retain_length = int(len(graded)*self.retain)

        # The parents are every network we want to keep.
        parents = graded[:retain_length]

        # For those we aren't keeping, randomly keep some anyway.
        for individual in graded[retain_length:]:
            if self.random_select > random.random():
                parents.append(individual)

        # Now find out how many spots we have left to fill.
        parents_length = len(parents)
        desired_length = len(pop) - parents_length
        children = []

        # Add children, which are bred from two remaining networks.
        while len(children) < desired_length:

            # Get a random mom and dad.
            print(parents_length-1)
            male = random.randint(0, parents_length-1)

            female = random.randint(0, parents_length-1)

            # Assuming they aren't the same network...
            if male != female:
                male = parents[male]
                female = parents[female]

                # Breed them.
                babies = self.breed(male, female)

                # Add the children one at a time.
                for baby in babies:
                    # Don't grow larger than desired length.
                    if len(children) < desired_length:
                        children.append(baby)

        parents.extend(children)

        return parents

In [31]:
import logging
from tqdm import tqdm

# Setup logging.
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filename='log.txt'
)

def train_networks(networks, dataset):
    """Train each network.
    Args:
        networks (list): Current population of networks
        dataset (str): Dataset to use for training/evaluating
    """

    pbar = tqdm(total=len(networks))
    logging.info("***a***")
    for network in networks:
        network.train(dataset)
        pbar.update(1)
        logging.info("***b***")
    pbar.close()

def get_average_accuracy(networks):
    """Get the average accuracy for a group of networks.
    Args:
        networks (list): List of networks
    Returns:
        float: The average accuracy of a population of networks.
    """
    total_accuracy = 0
    for network in networks:
        total_accuracy += network.accuracy

    return total_accuracy / len(networks)

def generate(generations, population, nn_param_choices, dataset):
    """Generate a network with the genetic algorithm.
    Args:
        generations (int): Number of times to evole the population
        population (int): Number of networks in each generation
        nn_param_choices (dict): Parameter choices for networks
        dataset (str): Dataset to use for training/evaluating
    """
    optimizer = Optimizer(nn_param_choices)
    logging.info("***Optimizer***")
    networks = optimizer.create_population(population)

    # Evolve the generation.
    for i in range(generations):
        logging.info("***Doing generation %d of %d***" %
                     (i + 1, generations))

        # Train and get accuracy for networks.
        train_networks(networks, dataset)
        logging.info("***Trained networks***")
        # Get the average accuracy for this generation.
        average_accuracy = get_average_accuracy(networks)

        # Print out the average accuracy each generation.
        logging.info("Generation average: %.2f%%" % (average_accuracy * 100))
        logging.info('-'*80)

        # Evolve, except on the last iteration.
        if i != generations - 1:
            # Do the evolution.
            networks = optimizer.evolve(networks)

    # Sort our final population.
    networks = sorted(networks, key=lambda x: x.accuracy, reverse=True)

    # Print out the top 5 networks.
    print_networks(networks[:5])

def print_networks(networks):
    """Print a list of networks.
    Args:
        networks (list): The population of networks
    """
    logging.info('-'*80)
    for network in networks:
        network.print_network()

def main():
    """Evolve a network."""
    generations = 10 # Number of times to evole the population.
    population = 20 # Number of networks in each generation.
    dataset = 'mnist'
    print(2)

    nn_param_choices = {
        #'nb_neurons': [10, 60, 110, 160, 210, 260,310,360,410,460,510],
        'VGG_layers': range(0,5),
        'VGG_sublayers': range(1,4),
        'pool_num' : [2],
        'nb_neurons1': range(14,780),
        'nb_neurons2':  range(13,779),
        'nb_neurons3': range(12,778),
        'nb_neurons4': range(11,777),
        'nb_layers': [1, 2, 3, 4],
        'activation': ['relu', 'elu', 'tanh', 'sigmoid'],
        'optimizer': ['rmsprop', 'adam', 'sgd', 'adagrad',
                      'adadelta', 'adamax', 'nadam'],
    }
    print(3)
    logging.info("***Evolving %d generations with population %d***" %
                 (generations, population))
    print(4)
    generate(generations, population, nn_param_choices, dataset)
    print('done')

if __name__ == '__main__':
    main()





  0%|          | 0/20 [00:00<?, ?it/s]

2
3
4
done mnist
compiled model






  5%|▌         | 1/20 [00:30<09:48, 30.98s/it]

0.7437999844551086
done mnist
compiled model






 10%|█         | 2/20 [01:16<10:37, 35.43s/it]

0.7447999715805054
done mnist
compiled model






 15%|█▌        | 3/20 [05:21<27:48, 98.16s/it]

0.4320000112056732
done mnist
compiled model






 20%|██        | 4/20 [06:31<23:56, 89.79s/it]

0.7534000277519226
done mnist
compiled model






 25%|██▌       | 5/20 [14:34<51:56, 207.80s/it]

0.33649998903274536
done mnist
compiled model






 30%|███       | 6/20 [21:28<1:02:53, 269.52s/it]

0.6791999936103821
done mnist
compiled model






 35%|███▌      | 7/20 [22:31<44:57, 207.53s/it]  

0.7240999937057495
done mnist
compiled model






 40%|████      | 8/20 [23:02<30:55, 154.64s/it]

0.10000000149011612
done mnist
compiled model






 45%|████▌     | 9/20 [28:02<36:21, 198.30s/it]

0.6972000002861023
done mnist
compiled model






 50%|█████     | 10/20 [29:04<26:13, 157.34s/it]

0.4083999991416931
done mnist
compiled model






 55%|█████▌    | 11/20 [32:31<25:51, 172.43s/it]

0.40220001339912415
done mnist
compiled model






 60%|██████    | 12/20 [35:05<22:15, 166.89s/it]

0.48069998621940613
done mnist
compiled model






 65%|██████▌   | 13/20 [39:53<23:40, 202.95s/it]

0.10000000149011612
done mnist
compiled model






 70%|███████   | 14/20 [41:13<16:37, 166.25s/it]

0.2964000105857849
done mnist
compiled model






 75%|███████▌  | 15/20 [45:50<16:36, 199.39s/it]

0.5637000203132629
done mnist
compiled model






 80%|████████  | 16/20 [46:21<09:55, 148.77s/it]

0.10000000149011612
done mnist
compiled model






 85%|████████▌ | 17/20 [47:41<06:24, 128.31s/it]

0.6852999925613403
done mnist
compiled model






 90%|█████████ | 18/20 [48:30<03:29, 104.63s/it]

0.4918999969959259
done mnist
compiled model






 95%|█████████▌| 19/20 [53:19<02:39, 159.71s/it]

0.43540000915527344
done mnist
compiled model






100%|██████████| 20/20 [54:28<00:00, 163.43s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7716000080108643
7
7
7
7
7
7
7
done mnist
compiled model






 45%|████▌     | 9/20 [00:48<00:59,  5.37s/it]

0.7558000087738037
done mnist
compiled model






 50%|█████     | 10/20 [02:09<04:40, 28.09s/it]

0.728600025177002
done mnist
compiled model






 55%|█████▌    | 11/20 [02:59<05:11, 34.67s/it]

0.7439000010490417
done mnist
compiled model






 60%|██████    | 12/20 [05:33<09:24, 70.53s/it]

0.4163999855518341
done mnist
compiled model






 65%|██████▌   | 13/20 [06:19<07:22, 63.23s/it]

0.7213000059127808
done mnist
compiled model






 70%|███████   | 14/20 [06:59<05:36, 56.08s/it]

0.7214000225067139
done mnist
compiled model






 75%|███████▌  | 15/20 [07:34<04:08, 49.68s/it]

0.10000000149011612
done mnist
compiled model






 80%|████████  | 16/20 [11:30<07:02, 105.60s/it]

0.6937000155448914
done mnist
compiled model






 85%|████████▌ | 17/20 [12:58<05:01, 100.43s/it]

0.7146000266075134
done mnist
compiled model






 90%|█████████ | 18/20 [14:49<03:27, 103.67s/it]

0.7146000266075134
done mnist
compiled model






 95%|█████████▌| 19/20 [15:20<01:21, 81.94s/it] 

0.6923999786376953
done mnist
compiled model






100%|██████████| 20/20 [16:52<00:00, 50.63s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.6951000094413757
7
7
7
7
7
7
7
7
done mnist
compiled model






 45%|████▌     | 9/20 [01:21<01:40,  9.11s/it]

0.7577000260353088
done mnist
compiled model






 50%|█████     | 10/20 [02:22<04:05, 24.52s/it]

0.7418000102043152
done mnist
compiled model






 55%|█████▌    | 11/20 [03:54<06:43, 44.79s/it]

0.7670999765396118
done mnist
compiled model






 60%|██████    | 12/20 [04:42<06:05, 45.70s/it]

0.741100013256073
done mnist
compiled model






 65%|██████▌   | 13/20 [05:43<05:51, 50.27s/it]

0.758400022983551
done mnist
compiled model






 70%|███████   | 14/20 [06:42<05:17, 52.96s/it]

0.10000000149011612
done mnist
compiled model






 75%|███████▌  | 15/20 [08:52<06:20, 76.06s/it]

0.44369998574256897
done mnist
compiled model






 80%|████████  | 16/20 [09:59<04:53, 73.47s/it]

0.7364000082015991
done mnist
compiled model






 85%|████████▌ | 17/20 [11:00<03:29, 69.72s/it]

0.7627999782562256
done mnist
compiled model






 90%|█████████ | 18/20 [11:49<02:06, 63.32s/it]

0.7319999933242798
done mnist
compiled model






 95%|█████████▌| 19/20 [12:56<01:04, 64.44s/it]

0.746399998664856
done mnist
compiled model






100%|██████████| 20/20 [13:39<00:00, 40.96s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7368999719619751
8
8
8
8
8
8
done mnist
compiled model






 50%|█████     | 10/20 [01:23<01:23,  8.34s/it]

0.7297999858856201
done mnist
compiled model






 55%|█████▌    | 11/20 [02:19<03:24, 22.71s/it]

0.7541999816894531
done mnist
compiled model






 60%|██████    | 12/20 [03:12<04:13, 31.68s/it]

0.7491999864578247
done mnist
compiled model






 65%|██████▌   | 13/20 [04:30<05:19, 45.66s/it]

0.10000000149011612
done mnist
compiled model






 70%|███████   | 14/20 [05:33<05:05, 50.86s/it]

0.751800000667572
done mnist
compiled model






 75%|███████▌  | 15/20 [07:29<05:51, 70.39s/it]

0.7699000239372253
done mnist
compiled model






 80%|████████  | 16/20 [09:19<05:29, 82.43s/it]

0.4884999990463257
done mnist
compiled model






 85%|████████▌ | 17/20 [10:29<03:55, 78.50s/it]

0.7455999851226807
done mnist
compiled model






 90%|█████████ | 18/20 [12:07<02:49, 84.52s/it]

0.7578999996185303
done mnist
compiled model






 95%|█████████▌| 19/20 [13:11<01:18, 78.14s/it]

0.7540000081062317
done mnist
compiled model






100%|██████████| 20/20 [14:11<00:00, 42.59s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7501000165939331
7
7
7
7
7
7
7
done mnist
compiled model






 45%|████▌     | 9/20 [01:06<01:21,  7.40s/it]

0.7506999969482422
done mnist
compiled model






 50%|█████     | 10/20 [02:10<04:03, 24.39s/it]

0.701200008392334
done mnist
compiled model






 55%|█████▌    | 11/20 [02:45<04:09, 27.68s/it]

0.10000000149011612
done mnist
compiled model






 60%|██████    | 12/20 [03:50<05:09, 38.68s/it]

0.7261999845504761
done mnist
compiled model






 65%|██████▌   | 13/20 [04:53<05:22, 46.14s/it]

0.7595999836921692
done mnist
compiled model






 70%|███████   | 14/20 [06:29<06:06, 61.11s/it]

0.765500009059906
done mnist
compiled model






 75%|███████▌  | 15/20 [10:37<09:44, 116.95s/it]

0.6316999793052673
done mnist
compiled model






 80%|████████  | 16/20 [11:38<06:40, 100.22s/it]

0.7430999875068665
done mnist
compiled model






 85%|████████▌ | 17/20 [12:43<04:28, 89.66s/it] 

0.7444000244140625
done mnist
compiled model






 90%|█████████ | 18/20 [13:47<02:43, 81.96s/it]

0.10000000149011612
done mnist
compiled model






 95%|█████████▌| 19/20 [15:49<01:33, 93.88s/it]

0.7663999795913696
done mnist
compiled model






100%|██████████| 20/20 [16:47<00:00, 50.37s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7328000068664551
7
7
7
7
7
7
done mnist
compiled model






 45%|████▌     | 9/20 [01:49<02:13, 12.17s/it]

0.7315000295639038
done mnist
compiled model






 50%|█████     | 10/20 [04:35<09:44, 58.40s/it]

0.4975000023841858
done mnist
compiled model






 55%|█████▌    | 11/20 [05:49<09:26, 62.96s/it]

0.7099999785423279
done mnist
compiled model






 60%|██████    | 12/20 [07:42<10:23, 77.94s/it]

0.7681000232696533
done mnist
compiled model






 65%|██████▌   | 13/20 [08:43<08:30, 72.94s/it]

0.767799973487854
done mnist
compiled model






 70%|███████   | 14/20 [10:02<07:28, 74.71s/it]

0.7531999945640564
done mnist
compiled model






 75%|███████▌  | 15/20 [10:48<05:31, 66.21s/it]

0.7419999837875366
done mnist
compiled model






 80%|████████  | 16/20 [12:55<05:37, 84.37s/it]

0.7602999806404114
done mnist
compiled model






 85%|████████▌ | 17/20 [14:08<04:02, 80.91s/it]

0.4803999960422516
done mnist
compiled model






 90%|█████████ | 18/20 [15:50<02:54, 87.21s/it]

0.7567999958992004
done mnist
compiled model






 95%|█████████▌| 19/20 [17:30<01:31, 91.05s/it]

0.7678999900817871
done mnist
compiled model






100%|██████████| 20/20 [19:06<00:00, 57.31s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7709000110626221
9
9
9
9
9
done mnist
compiled model






 55%|█████▌    | 11/20 [00:47<00:39,  4.36s/it]

0.7516000270843506
done mnist
compiled model






 60%|██████    | 12/20 [01:40<02:31, 18.89s/it]

0.7508000135421753
done mnist
compiled model






 65%|██████▌   | 13/20 [03:42<05:47, 49.63s/it]

0.7544000148773193
done mnist
compiled model






 70%|███████   | 14/20 [05:18<06:21, 63.53s/it]

0.7409999966621399
done mnist
compiled model






 75%|███████▌  | 15/20 [07:02<06:19, 75.80s/it]

0.743399977684021
done mnist
compiled model






 80%|████████  | 16/20 [08:48<05:39, 84.88s/it]

0.753000020980835
done mnist
compiled model






 85%|████████▌ | 17/20 [10:08<04:10, 83.40s/it]

0.7706000208854675
done mnist
compiled model






 90%|█████████ | 18/20 [11:29<02:45, 82.66s/it]

0.7486000061035156
done mnist
compiled model






 95%|█████████▌| 19/20 [13:09<01:27, 87.80s/it]

0.7436000108718872
done mnist
compiled model






100%|██████████| 20/20 [14:34<00:00, 43.75s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7649999856948853
9
9
9
9
9
done mnist
compiled model






 55%|█████▌    | 11/20 [01:01<00:50,  5.58s/it]

0.7657999992370605
done mnist
compiled model






 60%|██████    | 12/20 [02:50<04:52, 36.55s/it]

0.7766000032424927
done mnist
compiled model






 65%|██████▌   | 13/20 [04:34<06:37, 56.81s/it]

0.7746000289916992
done mnist
compiled model






 70%|███████   | 14/20 [06:11<06:53, 68.94s/it]

0.7603999972343445
done mnist
compiled model






 75%|███████▌  | 15/20 [07:01<05:16, 63.36s/it]

0.7074000239372253
done mnist
compiled model






 80%|████████  | 16/20 [08:02<04:10, 62.65s/it]

0.7674999833106995
done mnist
compiled model






 85%|████████▌ | 17/20 [10:04<04:00, 80.27s/it]

0.7662000060081482
done mnist
compiled model






 90%|█████████ | 18/20 [11:23<02:40, 80.07s/it]

0.7623999714851379
done mnist
compiled model






 95%|█████████▌| 19/20 [12:20<01:13, 73.03s/it]

0.7513999938964844
done mnist
compiled model






100%|██████████| 20/20 [12:57<00:00, 38.87s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.6970999836921692
10
10
10
10
10
10
done mnist
compiled model






 60%|██████    | 12/20 [00:36<00:24,  3.07s/it]

0.703000009059906
done mnist
compiled model






 65%|██████▌   | 13/20 [01:31<02:10, 18.63s/it]

0.10000000149011612
done mnist
compiled model






 70%|███████   | 14/20 [02:28<03:00, 30.06s/it]

0.7440000176429749
done mnist
compiled model






 75%|███████▌  | 15/20 [03:46<03:42, 44.57s/it]

0.7605999708175659
done mnist
compiled model






 80%|████████  | 16/20 [04:46<03:16, 49.21s/it]

0.677299976348877
done mnist
compiled model






 85%|████████▌ | 17/20 [05:52<02:42, 54.06s/it]

0.7595999836921692
done mnist
compiled model






 90%|█████████ | 18/20 [07:10<02:02, 61.20s/it]

0.7663999795913696
done mnist
compiled model






 95%|█████████▌| 19/20 [08:15<01:02, 62.43s/it]

0.7656000256538391
done mnist
compiled model






100%|██████████| 20/20 [10:04<00:00, 30.21s/it]




  0%|          | 0/20 [00:00<?, ?it/s]

0.7627000212669373
7
7
7
7
7
7
done mnist
compiled model






 45%|████▌     | 9/20 [01:05<01:19,  7.26s/it]

0.769599974155426
done mnist
compiled model






 50%|█████     | 10/20 [02:41<05:37, 33.80s/it]

0.7641000151634216
done mnist
compiled model






 55%|█████▌    | 11/20 [03:41<06:15, 41.70s/it]

0.7645999789237976
done mnist
compiled model






 60%|██████    | 12/20 [05:28<08:11, 61.41s/it]

0.7551000118255615
done mnist
compiled model






 65%|██████▌   | 13/20 [07:25<09:06, 78.14s/it]

0.770799994468689
done mnist
compiled model






 70%|███████   | 14/20 [10:24<10:50, 108.35s/it]

0.6995000243186951
done mnist
compiled model






 75%|███████▌  | 15/20 [12:00<08:42, 104.55s/it]

0.7634999752044678
done mnist
compiled model






 80%|████████  | 16/20 [13:29<06:39, 99.85s/it] 

0.760699987411499
done mnist
compiled model






 85%|████████▌ | 17/20 [15:01<04:53, 97.70s/it]

0.7494000196456909
done mnist
compiled model






 90%|█████████ | 18/20 [16:55<03:24, 102.48s/it]

0.7544000148773193
done mnist
compiled model






 95%|█████████▌| 19/20 [18:40<01:43, 103.22s/it]

0.7587000131607056
done mnist
compiled model






100%|██████████| 20/20 [20:16<00:00, 60.84s/it] 

0.758899986743927
done


04/12/2020 09:21:47 PM - INFO - {'VGG_layers': 4, 'VGG_sublayers': 3, 'pool_num': 2, 'nb_neurons1': 248, 'nb_neurons2': 35, 'nb_neurons3': 17, 'nb_neurons4': 11, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'adamax'}
04/12/2020 09:21:47 PM - INFO - Network accuracy: 77.66%
04/12/2020 09:21:47 PM - INFO - {'VGG_layers': 4, 'VGG_sublayers': 3, 'pool_num': 2, 'nb_neurons1': 166, 'nb_neurons2': 35, 'nb_neurons3': 17, 'nb_neurons4': 11, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'adamax'}
04/12/2020 09:21:47 PM - INFO - Network accuracy: 77.46%
04/12/2020 09:21:47 PM - INFO - {'VGG_layers': 4, 'VGG_sublayers': 3, 'pool_num': 2, 'nb_neurons1': 29, 'nb_neurons2': 26, 'nb_neurons3': 17, 'nb_neurons4': 13, 'nb_layers': 3, 'activation': 'elu', 'optimizer': 'rmsprop'}
04/12/2020 09:21:47 PM - INFO - Network accuracy: 77.16%
04/12/2020 09:21:47 PM - INFO - {'VGG_layers': 4, 'VGG_sublayers': 3, 'pool_num': 2, 'nb_neurons1': 248, 'nb_neurons2': 170, 'nb_neurons3': 138, 'nb_neurons4': 11, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'adamax'}
04/12/2020 09:21:47 PM - INFO - Network accuracy: 77.09%
04/12/2020 09:21:47 PM - INFO - {'VGG_layers': 4, 'VGG_sublayers': 3, 'pool_num': 2, 'nb_neurons1': 298, 'nb_neurons2': 26, 'nb_neurons3': 17, 'nb_neurons4': 11, 'nb_layers': 3, 'activation': 'relu', 'optimizer': 'adamax'}
04/12/2020 09:21:47 PM - INFO - Network accuracy: 77.08%


Conclusion, my gentic code is good for MLPS or for when you already have a rough idea of how the network will look, so just for optimising parameters.